### PR TITLE
EOS-16272: cfgen to generate aux pool layouts considering node failures

### PR DESCRIPTION
**Pull request by suvratjoshi**
This feature is developed considering the node (enclosure) failure
case. If a node in the cluster fails, it means that all the disks
in that node become inaccessible. As this feature considers only
node failures, it keeps the spare pools ready considering the
other alive nodes from the cluster.
The feature works using the mathematical combinations formula
C(n, r) = (n! / (r!(n-r)!))
It basically creates combinations of all the nodes in case of
failures of nodes from 1 to the allowed number of nodes to be
failed, by allowance vector.
For each and every such generated combinations of (alive) nodes,
it creates the sets of the disks and assigns a pool to each of
them. Such spare pools created considering allowed node failures
are called as aux pools.

This PR also covers https://github.com/Seagate/cortx-hare/issues/1742

Signed-off-by: Suvrat Joshi <suvrat.joshi@seagate.com>

### CODE REVIEW

File: cfgen/cfgen [hunk 1]
```diff
@@ -1854,7 +1858,38 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
                 proc_id = ConfProcess.build(conf, node_id, node, proc_t)
                 cluster.m0_clients[proc_id] = new_oid(ObjT.service)
 
+    aux_pools = {}
+    disk_ref_combinations = []
     for pool in cluster_desc['pools']:
+        parity_group_size = pool['data_units'] + pool['parity_units'] +\
+                            pool['spare_units']
+        if len(pool.get('disk_refs', [])) == parity_group_size:
```
Comment (suvratjoshi) [inline cfgen/cfgen:1866]: @mssawant 
This is that strict condition of disks to be equal.
Once you check code / conditions and confirm, I will remove this.
Thanks!

**Review by mssawant — COMMENTED** [review]
Also I tried your branch and I did manage to generate multiple aux pools (by having pool width == N + K + S) but hitting this crash,
```
(gdb) bt
#0  0x00007f71bd481387 in raise () from /lib64/libc.so.6
#1  0x00007f71bd482a78 in abort () from /lib64/libc.so.6
#2  0x00007f71bef189dd in m0_arch_panic (c=c@entry=0x7f71bf36e800 <__pctx.14230>, ap=ap@entry=0x7ffe0984ae98) at lib/user_space/uassert.c:131
#3  0x00007f71bef08a64 in m0_panic (ctx=ctx@entry=0x7f71bf36e800 <__pctx.14230>) at lib/assert.c:52
#4  0x00007f71beeca1ac in m0_fd__tile_init (tile=tile@entry=0x2540868, la_attr=la_attr@entry=0x24184b8, children=children@entry=0x7ffe0984b060, depth=3) at fd/fd.c:221
#5  0x00007f71beeca511 in m0_fd_tile_build (pv=pv@entry=0x24183b0, pool_ver=pool_ver@entry=0x25407c0, failure_level=failure_level@entry=0x7ffe0984b110) at fd/fd.c:281
#6  0x00007f71bef70158 in m0_pool_version_init_by_conf (pv=pv@entry=0x25407c0, pver=pver@entry=0x24183b0, pool=pool@entry=0x2417240, pc=pc@entry=0x7ffe0984c380) at pool/pool.c:769
#7  0x00007f71bef71b2d in m0_pool_versions_setup (pc=pc@entry=0x7ffe0984c380) at pool/pool.c:1674
#8  0x00007f71bef291ac in cs_level_enter (module=<optimized out>) at motr/setup.c:2683
#9  0x00007f71bef594ce in module_up (module=module@entry=0x7ffe09859c20, level=level@entry=42) at module/module.c:121
#10 0x00007f71bef5956a in m0_module_init (module=module@entry=0x7ffe09859c20, level=level@entry=42) at module/module.c:137
#11 0x00007f71bef2adc1 in m0_cs_start (cctx=cctx@entry=0x7ffe0984c380) at motr/setup.c:2946
#12 0x0000000000401207 in main (argc=24, argv=0x7ffe0985aea8) at motr/m0d.c:198
```
This is because the pool width > the tile width i.e. the number of devices are less than the pool width but in actual the pool does have 4 devices ( equal to pool width) but again it is the problem of symmetric tree. As one node has 2 devices and just 1 device from rest of the 2 nodes.
We need to try with a large pool width by removing the pool width restriction.

File: cfgen/cfgen [hunk 2]
```diff
@@ -1182,14 +1182,18 @@ def pool_drives(m0conf: Dict[Oid, Any],
...
-    if over_referenced:
-        err = '{} referred to by several pools:'.format(
-            'This disk is' if len(over_referenced) == 1 else 'These disks are')
-        for node_id in sorted(over_referenced):
-            err += '\n  ' + m0conf[node_id].name
-            for drive_id in sorted(over_referenced[node_id]):
-                err += '\n   \\_ ' + m0conf[m0conf[drive_id].sdev].filename
-        raise AssertionError(err)
+
+    # TODO: coordinate with Motr component, to make sure that it supports the
```
[the remark this answers is not in the document]
Comment (suvratjoshi) [reply cfgen/cfgen]: Yes.
On Cortx.Hare teams channel, this was [confirmed](https://teams.microsoft.com/l/message/19:af82b72790004dff885bd374a2fda099@thread.tacv2/1627036476286?tenantId=d466216a-c643-434a-9c2e-057448c17cbe&groupId=7e2a2869-5847-4a65-9402-f4778c7b913e&parentMessageId=1619181794073&teamName=CORTX.All&channelName=CORTX.Hare&createdTime=1627036476286) with Madhav.
```
[7/23 8:31 AM] Suvrat Joshi Chandrashekhar Joshi
    Hi Madhavrao,
For AUX pools feature, I would like to confirm if Motr : supports the same storage device to be part of more than one pool.
Can you please confirm this?
Thanks!

​[7/23 4:04 PM] Madhavrao Vemuri
    Yes, devices can be under different aux. pools.
```

File: cfgen/cfgen (hunk 1 above)
Comment (suvratjoshi) [reply cfgen/cfgen]: Removed.
Algorithm is now completely changed from the original one.
Original algo was creating combinations considering tolerance values of disks.
New algo creates node combinations considering tolerance values for node first and then creates lists of corresponding disks, accordingly.

File: cfgen/cfgen [hunk 3]
```diff
@@ -1182,14 +1182,18 @@ def pool_drives(m0conf: Dict[Oid, Any],
...
+    #       same storage device to be part of more than one pool
+
+    # if over_referenced:
```
[the remark this answers is not in the document]
Comment (suvratjoshi) [reply cfgen/cfgen]: Done!

File: cfgen/cfgen [hunk 4]
```diff
@@ -1854,7 +1858,38 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
                 proc_id = ConfProcess.build(conf, node_id, node, proc_t)
                 cluster.m0_clients[proc_id] = new_oid(ObjT.service)
 
+    aux_pools = {}
+    disk_ref_combinations = []
     for pool in cluster_desc['pools']:
+        parity_group_size = pool['data_units'] + pool['parity_units'] +\
```
[the remark this answers is not in the document]
Comment (suvratjoshi) [reply cfgen/cfgen]: Had written this function but later changed the algorithm.
So not needed now.

File: cfgen/cfgen (hunk 1 above)
[the remark this answers is not in the document]
Comment (suvratjoshi) [reply cfgen/cfgen]: Removed.
Algorithm is now completely changed from the original one.
Original algo was creating combinations considering tolerance values of disks.
New algo creates node combinations considering tolerance values for node first and then creates lists of corresponding disks, accordingly.

**Comment by suvratjoshi** [comment]
Example bootstrap of a three node cluster with layout parameters :
```
pools:
  - name: the pool
    disk_refs:
      - { path: /dev/sdb, node: ssc-vm-2493.colo.seagate.com }
      - { path: /dev/sdc, node: ssc-vm-2493.colo.seagate.com }
      - { path: /dev/sdb, node: ssc-vm-6328.colo.seagate.com }
      - { path: /dev/sdc, node: ssc-vm-6328.colo.seagate.com }
      - { path: /dev/sdb, node: ssc-vm-6327.colo.seagate.com }
      - { path: /dev/sdc, node: ssc-vm-6327.colo.seagate.com }
    type: sns  # optional; supported values: "sns" (default), "dix", "md"
    data_units: 4
    parity_units: 2
    spare_units: 0
```

Bootstrap is successful :
```
[root@ssc-vm-2493 fin_test]# hctl bootstrap --mkfs three_node.yaml
2021-08-19 05:10:47: Generating cluster configuration... OK
2021-08-19 05:10:51: Starting Consul server on this node............ OK
2021-08-19 05:11:01: Importing configuration into the KV store... OK
2021-08-19 05:11:01: Starting Consul on other nodes...Consul ready on all nodes
2021-08-19 05:11:03: Updating Consul configuraton from the KV store... OK
2021-08-19 05:11:05: Waiting for the RC Leader to get elected... OK
2021-08-19 05:11:05: Starting Motr (phase1, mkfs)... OK
2021-08-19 05:11:13: Starting Motr (phase1, m0d)... OK
2021-08-19 05:11:17: Starting Motr (phase2, mkfs)... OK
2021-08-19 05:11:26: Starting Motr (phase2, m0d)... OK
2021-08-19 05:11:31: Checking health of services... OK
[root@ssc-vm-2493 fin_test]#
```

Here we can see in the status that aux pools are created.
```
[root@ssc-vm-2493 fin_test]# hctl status -d
Data pools:
    # fid name
    0x6f00000000000001:0x57 'the pool'
    0x6f00000000000001:0x72 'the pool-aux001'
    0x6f00000000000001:0x89 'the pool-aux002'
    0x6f00000000000001:0xa0 'the pool-aux003'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xd7 'default': 'the pool' 'the pool-aux001' 'the pool-aux002' 'the pool-aux003' None None
Services:
    ssc-vm-6328.colo.seagate.com
    [started]  hax        0x7200000000000001:0x22  192.168.47.140@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x25  192.168.47.140@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0x28  192.168.47.140@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x35  192.168.47.140@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x38  192.168.47.140@tcp:12345:4:2
    ssc-vm-2493.colo.seagate.com
    [started]  hax        0x7200000000000001:0x6   192.168.52.140@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x9   192.168.52.140@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0xc   192.168.52.140@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x19  192.168.52.140@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x1c  192.168.52.140@tcp:12345:4:2
    ssc-vm-6327.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x3e  192.168.47.126@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x41  192.168.47.126@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0x44  192.168.47.126@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x51  192.168.47.126@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x54  192.168.47.126@tcp:12345:4:2
Devices:
    ssc-vm-6328.colo.seagate.com
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/null
    ssc-vm-2493.colo.seagate.com
    [online]  /dev/null
    [online]  /dev/sdc
    [online]  /dev/sdb
    ssc-vm-6327.colo.seagate.com
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/null
[root@ssc-vm-2493 fin_test]#
```

consul kv get --recurse | grep aux
```
m0conf/pools/0x6f00000000000001:0x72:the pool-aux001
m0conf/pools/0x6f00000000000001:0x89:the pool-aux002
m0conf/pools/0x6f00000000000001:0xa0:the pool-aux003
m0conf/profiles/0x7000000000000001:0xd7:{"name": "default", "pools": ["the pool", "the pool-aux001", "the pool-aux002", "the pool-aux003", null, null]}
```

cat /var/lib/hare/confd.dhall | grep tolerance
```
  , types.Obj.Pver { id = utils.zoid types.ObjT.Pver 88, data_units = 4, parity_units = 2, spare_units = 0, pool_width = 6, tolerance = [0, 0, 1, 1, 2], sitevs = [ utils.zoid types.ObjT.Objv 93 ] }
  , types.Obj.Pver { id = utils.zoid types.ObjT.Pver 115, data_units = 2, parity_units = 2, spare_units = 0, pool_width = 4, tolerance = [0, 0, 1, 1, 2], sitevs = [ utils.zoid types.ObjT.Objv 120 ] }
  , types.Obj.Pver { id = utils.zoid types.ObjT.Pver 138, data_units = 2, parity_units = 2, spare_units = 0, pool_width = 4, tolerance = [0, 0, 1, 1, 2], sitevs = [ utils.zoid types.ObjT.Objv 143 ] }
  , types.Obj.Pver { id = utils.zoid types.ObjT.Pver 161, data_units = 2, parity_units = 2, spare_units = 0, pool_width = 4, tolerance = [0, 0, 1, 1, 2], sitevs = [ utils.zoid types.ObjT.Objv 166 ] }
  , types.Obj.Pver { id = utils.zoid types.ObjT.Pver 190, data_units = 1, parity_units = 2, spare_units = 0, pool_width = 3, tolerance = [0, 0, 2, 2, 2], sitevs = [ utils.zoid types.ObjT.Objv 195 ] }
  , types.Obj.Pver { id = utils.zoid types.ObjT.Pver 203, data_units = 1, parity_units = 2, spare_units = 0, pool_width = 3, tolerance = [0, 0, 2, 2, 2], sitevs = [ utils.zoid types.ObjT.Objv 208 ] }
```

**Comment by suvratjoshi** [comment]
In this PR, I am also going to add changes for https://github.com/Seagate/cortx-hare/issues/1742.

File: cfgen/cfgen [hunk 5]
```diff
@@ -1956,8 +1948,66 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
...
+                                     d['ctrl'], d['disk'])
+            else:
+                tolerance = ConfPool.gen_sns_tolerance(conf, pool)
+            # Generate the node combinations considering node failures as
+            # allowed by node tolerance values
+            temp_pool_nodes = []
+            for nodes in pool['disk_refs']:
+                temp_pool_nodes += [
+                    nodes['node']]
+            pool_nodes = list(set(temp_pool_nodes))
```
Comment (knekrasov) [inline cfgen/cfgen:1976]: Is it just a way to guarantee the uniqueness in poll_nodes?

File: cfgen/cfgen [hunk 6]
```diff
@@ -1956,8 +1948,66 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
...
+    disk_ref_combinations = []
+    encl_ref_combination = []
+    tolerance: Failures
+    pool_nodes: list = []
+
     for pool in cluster_desc['pools']:
-        ConfPool.build(conf, root_id, pool)
+        pool_t = pool_type(pool)
+        if pool_t == PoolT.sns:
+            # Calculate and create aux pool disk combinations
```
Comment (knekrasov) [inline cfgen/cfgen:1961]: [Optional] Can this `if` branch be refactored into a separate function?

File: cfgen/cfgen [hunk 7]
```diff
@@ -1956,8 +1948,66 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
...
+                tolerance = ConfPool.gen_sns_tolerance(conf, pool)
+            # Generate the node combinations considering node failures as
+            # allowed by node tolerance values
+            temp_pool_nodes = []
+            for nodes in pool['disk_refs']:
+                temp_pool_nodes += [
+                    nodes['node']]
+            pool_nodes = list(set(temp_pool_nodes))
+            for encl_failures in range(tolerance.enclosure, 0, -1):
+                encl_ref_combination += [
```
Comment (knekrasov) [inline cfgen/cfgen:1978]: Can you use type hints in your variables? 

I can see that the overall algorithm is a bit tricky, there are no unit tests. So having at least the type hints could highlight the intentions and involve static analysis into the game.

File: cfgen/cfgen (hunk 5 above)
Comment (suvratjoshi) [reply cfgen/cfgen:1976]: I guess not. There are multiple ways to find unique ones.
But I felt this is simple technique to get only unique ones in the list.

File: cfgen/cfgen (hunk 6 above)
Comment (suvratjoshi) [reply cfgen/cfgen]: Done!

File: cfgen/cfgen (hunk 7 above)
Comment (suvratjoshi) [reply cfgen/cfgen]: @knekrasov 
Done!
Thank you!

**Comment by suvratjoshi** [comment]
Testing done : 
1. One node : Bootstrap, hctl status, m0cp write and m0cat read
2. Three node : Bootstrap, hctl status, m0cp write and m0cat read
3. Three node with dix pool: Bootstrap, hctl status, m0cp write and m0cat read
4. Six node : Bootstrap, hctl status, m0cp write and m0cat read

Testing is done for pools with symmetric disk structure only (all nodes having same number of disks)
Tolerance values passed through CDF and also allowed to get generated without passing through CDF.
All above cases worked fine.

Custom build created successfully :
http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.8/job/cortx-custom-ci/2560/
RPMs at location:
http://cortx-storage.colo.seagate.com/releases/cortx/github/integration-custom-ci/centos-7.8.2003/custom-build-2560

3-Node deployment successful :
http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-R2-3Node/781/

**Review by knekrasov — APPROVED** [review]

**Review by mssawant — APPROVED** [review]

**Pull request merged**